### ISSUE TITLE
Support Ractors as Cluster Workers in addition to UNIX Processes

### ISSUE THREAD

**Issue by kigster** [puma/puma#3481]
**Is your feature request related to a problem? Please describe.**

Given that Ractors in Ruby 3 allow leveraging additional CPU cores. Yet, they share "shareable objects" such as `Class` and `Module` as well as all frozen constants; using Ractors as cluster workers should use less memory than the fully-fledged UNIX processes. 

This would allow a single Puma process never to have to `fork()` and yet be able to utilize all CPU cores.

**Describe the solution you'd like**

Perhaps instead of the configuration:

```ruby
workers Etc.nprocessors
```

The configuration DSL can support:

```ruby
# Previous behavior (the default)
workers Etc.nprocessors, type: :process

# New behavior on Ruby 3+
workers Etc.nprocessors, type: :ractor
```

**Describe alternatives you've considered**

In a cluster mode Puma already supports the pool of Process Workers. A decent chunk of Puma's code is handling worker processes via UNIX signals, restarting them, and so on.  

Since Ruby 3 Ractors can be used in a similar fashion, and can save quite a bit of memory (especially on very large Rails apps) and make Puma easier to deploy, manage and observe.

**Additional context**

https://blog.kiprosh.com/ruby-3-introduction-to-ractors/

**Comment by dentarg** [comment]
Some related existing issues/comments:

* https://github.com/puma/puma/issues/3422
* https://github.com/puma/puma/issues/2517#issuecomment-752231305

**Comment by dentarg** [comment]
I'm not sure this makes sense. The are no plans to make Rails compatible with Ractors: https://github.com/rails/rails/issues/51543#issuecomment-2050575058

Closing was wontfix for now.

**Comment by kigster** [comment]
This has nothing to do with rails. Each worker can run a Sinatra app or any other rack compliant service.

But using ractors you'd be saving on RAM, and yet utilizing multiple cores.

**Comment by dentarg** [comment]
Well, you mentioned Rails in your initial message, so I thought it was worth mentioning the practicality of it. You are of course right in that the framework  doesn't matter to Puma.

**Comment by dentarg** [comment]
Some recent related reading on this topic: https://byroot.github.io/ruby/performance/2025/02/27/whats-the-deal-with-ractors.html

**Comment by rgaufman** [comment]
The new RailsConf made a big deal about ractors, seems this is the primary answer to multi-core scaling with Ruby. Is this still something that's very much not planned, or has anything changed?

**Comment by joshuay03** [comment]
You can't simply swap processes with Ractors, as the latter can only share immutable objects, and this won't be true for the majority of applications. 

However, something that's stuck with me is @jhawthorn referring to Ractors as a no-GVL block - since each Ractor has its own GVL, they can execute CPU-intensive tasks in parallel. This means Puma could potentially experiment with offloading CPU-bound work like HTTP parsing to Ractors.
